#Exercise 1: Explore Ames housing dataset with descriptive statistics
In this exercise, we will learn how to explore a dataset and understand how it is structured. To do so, we will implement our own describe functions for numerical and categorical variables.
The dataset used for this exercise is the Ames Housing dataset compiled by Dean De Cock: http://www.amstat.org/publications/jse/v19n3/decock.pdf

1. Open on a new Colab notebook and import the pandas package

In [0]:
import pandas as pd

2. Assign the link to the AMES dataset to a variable called 'file_url':

In [0]:
file_url = 'https://raw.githubusercontent.com/TrainingByPackt/The-Data-Science-Workshop/master/Chapter10/dataset/ames_iowa_housing.csv'

3. Using the read_csv method from the package pandas, load the dataset into a new variable called 'df':

In [0]:
df = pd.read_csv(file_url)

4. Print the number of rows and columns of the dataframe using the attribute **shape** from **pandas** package:

Expected Output:

![Figure 20 - Number of rows and columns from the housing dataset](https://docs.google.com/uc?export=download&id=1r5wqla9_Sr4dk4KAd7wvqYo0rHexRTNT)

In [4]:
df.shape

(1460, 81)

We can see this dataset contains 1460 rows and 81 different columns. 

5. Print the names of the variables contained in this dataframe using the attribute **columns** from **pandas** package:

Expected Output:

![Figure 21 - List of columns in the housing dataset](https://docs.google.com/uc?export=download&id=1ISIHTmrLdSKHr6ZrCU0upxi149VLH6Vn)


In [5]:
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

We can infer the type of information contained in some of the variables by looking at their names such as LotArea (property size), YearBuilt (year of construction) or SalePrice (property sale price).

6. Print out the type of each variable contained in this dataframe using the attribute **dtypes** from **pandas** package:

Expected Output:

![Figure 22 - List of columns and their type from the housing dataset](https://docs.google.com/uc?export=download&id=1pL7D4IAk43QAL3ol2fsB7BzjQB81Yv33)


In [6]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
                  ...   
BedroomAbvGr       int64
KitchenAbvGr       int64
KitchenQual       object
TotRmsAbvGrd       int64
Functional        object
Fireplaces         int64
FireplaceQu       object
GarageType        object
GarageYrBlt      float64


We can see the variables are either numerical or text type. There is no date column in this dataset.


7. Display the first top rows of the dataframe using the method **head()** from **pandas**:

Expected Output:

![Figure 23 - First 5 rows of the housing dataset](https://docs.google.com/uc?export=download&id=1Oddvg6tvyy_NpwYvDy4R0JHROXtn5cdw)


In [7]:
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

8. Display the last bottom rows of the dataframe using the method **tail()** from **pandas**:

Expected Output:

![Figure 24 - Last 5 rows of the housing dataset](https://docs.google.com/uc?export=download&id=1qmil5wf0TCUTHg15VIRxv8KnR7uxHEJg)


In [8]:
df.tail()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
1455  1456          60       RL  ...        WD         Normal    175000
1456  1457          20       RL  ...        WD         Normal    210000
1457  1458          70       RL  ...        WD         Normal    266500
1458  1459          20       RL  ...        WD         Normal    142125
1459  1460          20       RL  ...        WD         Normal    147500

[5 rows x 81 columns]

It seems the column ‘Alley’ has a lot of missing values. The columns ‘Street’ and ‘Utilities’ seem to have only one value.

9. Display 5 random sampled rows of the dataframe using the method **sample()** from **pandas** and pass it a 'random_state' of 8:

Expected Output:

![Figure 25 - 5 random sampled rows of the housing dataset](https://docs.google.com/uc?export=download&id=1frsCDrsEEfWE-dkNUnHQdFm9iACWi1o0)


In [9]:
df.sample(n=5, random_state=8)

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
1260  1261          60       RL  ...        WD         Normal    181000
274    275          20       RL  ...        WD         Normal    124500
51      52          50       RM  ...        WD         Normal    114500
117    118          20       RL  ...       New        Partial    155000
789    790          60       RL  ...        WD         Normal    187500

[5 rows x 81 columns]

With this random samples, we can see that the column ‘LotFrontage’ also has some missing values. 

10. Create a new dataframe called 'num_df' with only the columns that are of numerical types using the method select_dtypes from pandas packages and pass in the value 'number' to the parameter 'include: 

In [0]:
num_df = df.select_dtypes(include='number')

11. Using the attribute **columns** from pandas, extract the list of columns of this dataframe 'num_df' and assign it to a new variable called 'num_cols' and print its content:

Expected Output:

![Figure 26 - List of numerical columns](https://docs.google.com/uc?export=download&id=1Xe6dR0fT6zT6-WIddLPvOOu2QU8OjMfX)



In [11]:
num_cols = num_df.columns
num_cols

Index(['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',
       'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

12. Perform the same steps as 7 and 8 but this time extract the list of columns that are of type 'object' and save it into a new variable called 'obj_cols':


Expected Output:

![Figure 27 - List of categorical variables](https://docs.google.com/uc?export=download&id=1n5gkiNhA0bPeZFUTdrXEssgqJtqwuBW0)


In [12]:
obj_cols = df.select_dtypes(include='object').columns
obj_cols

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')

13. Create a function called 'describe_object' that takes a **pandas** dataframe and a column name as input parameters. Then inside the function, print out the name of the given column, its number of unique values using the method **nunique()** and the list of values and their occurence using the method **value_counts()**:

In [0]:
def describe_object(df, col_name):
  print(f"\nCOLUMN: {col_name}")
  print(f"{df[col_name].nunique()} different values")
  print(f"List of values:")
  print(df[col_name].value_counts(dropna=False, normalize=True))

14. Test this function by providing the 'df' dataframe and the column 'MSZoning':

Expected Output:

![Figure 28 - Display of the created function for the column ‘MSZoning’](https://docs.google.com/uc?export=download&id=1PE8Ive1mpHZeeC0GN5wDrjxvkaufT5Pv)


In [14]:
describe_object(df, 'MSZoning')


COLUMN: MSZoning
5 different values
List of values:
RL         0.788356
RM         0.149315
FV         0.044521
RH         0.010959
C (all)    0.006849
Name: MSZoning, dtype: float64


For the column ‘MSZoning’, the value ‘RL’ represents almost 79% of the values while ‘C (all) is only present less than 1% of the rows.

15. Create a for loop that will call the created function for every element from the list 'obj_cols':

Expected Output:

![Figure 29 - Display of the created function for the first columns contained in ‘obj_cols’](https://docs.google.com/uc?export=download&id=1cou8u_aD4Z8ec1EqDKRPFXJGGyEQUzh_)



In [15]:
for col_name in obj_cols:
  describe_object(df, col_name)


COLUMN: MSZoning
5 different values
List of values:
RL         0.788356
RM         0.149315
FV         0.044521
RH         0.010959
C (all)    0.006849
Name: MSZoning, dtype: float64

COLUMN: Street
2 different values
List of values:
Pave    0.99589
Grvl    0.00411
Name: Street, dtype: float64

COLUMN: Alley
2 different values
List of values:
NaN     0.937671
Grvl    0.034247
Pave    0.028082
Name: Alley, dtype: float64

COLUMN: LotShape
4 different values
List of values:
Reg    0.633562
IR1    0.331507
IR2    0.028082
IR3    0.006849
Name: LotShape, dtype: float64

COLUMN: LandContour
4 different values
List of values:
Lvl    0.897945
Bnk    0.043151
HLS    0.034247
Low    0.024658
Name: LandContour, dtype: float64

COLUMN: Utilities
2 different values
List of values:
AllPub    0.999315
NoSeWa    0.000685
Name: Utilities, dtype: float64

COLUMN: LotConfig
5 different values
List of values:
Inside     0.720548
Corner     0.180137
CulDSac    0.064384
FR2        0.032192
FR3        0.00

We can confirm that the column ‘Street’ is almost constant as 99.6% of the rows contains the same value: ‘Pave’. For the column ‘Alley’, almost 94% of the rows have missing values.


16. Create a function called 'describe_numeric' that takes a **pandas** dataframe and a column name as input parameters. Then inside the function, print out the name of the given column, its minimum value using **min()**, maximum value using **max()**, its average value using **mean()**, standard deviation using **std()**, and median using **median()**:

In [0]:
def describe_numeric(df, col_name):
  print(f"\nCOLUMN: {col_name}")
  print(f"Minimum: {df[col_name].min()}")
  print(f"Maximum: {df[col_name].max()}")
  print(f"Average: {df[col_name].mean()}")
  print(f"Standard Deviation: {df[col_name].std()}")
  print(f"Median: {df[col_name].median()}")

17. Test this function by providing the 'df' dataframe and the column 'SalePrice':

Expected Output:

![Figure 30 - Display of the created function for the column ‘SalePrice’](https://docs.google.com/uc?export=download&id=1V0AZA0zxOcrbS2HLYFPhE75kBzy7CwPx)


In [17]:
describe_numeric(df, 'SalePrice')


COLUMN: SalePrice
Minimum: 34900
Maximum: 755000
Average: 180921.19589041095
Standard Deviation: 79442.50288288663
Median: 163000.0


The sale price ranges from 34,900 to 755,000 with an average of 180,921. The median is slightly lower than the average which tells us there are some outliers with high sales prices.

18. Create a for loop that will call the created function for every element from the list 'obj_cols':

Expected Output:

![Figure 31 - Display of the created function for the first columns contained in ‘num_cols’](https://docs.google.com/uc?export=download&id=1N9LXQhRwfC6HlILm1kJBMa3xwax_Sydp)

In [18]:
for col_name in num_cols:
  describe_numeric(df, col_name)


COLUMN: Id
Minimum: 1
Maximum: 1460
Average: 730.5
Standard Deviation: 421.6100093688479
Median: 730.5

COLUMN: MSSubClass
Minimum: 20
Maximum: 190
Average: 56.897260273972606
Standard Deviation: 42.30057099381035
Median: 50.0

COLUMN: LotFrontage
Minimum: 21.0
Maximum: 313.0
Average: 70.04995836802665
Standard Deviation: 24.284751774483183
Median: 69.0

COLUMN: LotArea
Minimum: 1300
Maximum: 215245
Average: 10516.828082191782
Standard Deviation: 9981.264932379147
Median: 9478.5

COLUMN: OverallQual
Minimum: 1
Maximum: 10
Average: 6.0993150684931505
Standard Deviation: 1.3829965467415923
Median: 6.0

COLUMN: OverallCond
Minimum: 1
Maximum: 9
Average: 5.575342465753424
Standard Deviation: 1.1127993367127367
Median: 5.0

COLUMN: YearBuilt
Minimum: 1872
Maximum: 2010
Average: 1971.267808219178
Standard Deviation: 30.202904042525265
Median: 1973.0

COLUMN: YearRemodAdd
Minimum: 1950
Maximum: 2010
Average: 1984.8657534246574
Standard Deviation: 20.645406807709396
Median: 1994.0

COLUMN: Ma

The column ‘Id’ ranges from 1 to 1460 which is the exact value as the number of rows in this dataset so this column is definitely a unique identifier of the property sold. It seems the values from the ‘MSSubClass’ are all rounded. This may indicates the information contained in this column has either been clustered into groups of ten or may be a categorical variable.

Congratulations on completing this exercise! We saw how to explore a newly received dataset with just a few lines of code. This helped us to understand its structure, the type of information contained in each variable and also identified some potential data quality issues such as missing values or incorrect values.